File: apps/schema_registry.py
import json
import urllib.request

SCHEMA_REGISTRY_URL = "http://schema-registry:8081"


def fetch_latest_schema(topic):
    """Fetch the current Avro value-schema for a Debezium CDC topic from the
    Schema Registry, as a JSON string for pyspark.sql.avro.functions.from_avro.

    Fetched once per entity at Spark job startup, not per-record. If the
    schema evolves while the job is running, messages written under the new
    schema either decode wrong/throw (incompatible change -- caught as a
    decode failure, see helper.parse_cdc_stream) or silently drop the new
    field (compatible/additive change) until the job is restarted to pick up
    the new schema -- a deliberate simplification over full per-record
    schema-ID resolution (e.g. ABRiS), accepted for this project's scale
    (PLAN.md 3.2).
    """
    subject = f"{topic}-value"
    url = f"{SCHEMA_REGISTRY_URL}/subjects/{subject}/versions/latest"
    with urllib.request.urlopen(url, timeout=10) as resp:
        body = json.loads(resp.read())
    return body["schema"]

File: scala-app/src/main/scala/pipeline/SchemaRegistry.scala
package pipeline

import java.net.URI
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.time.Duration

import com.fasterxml.jackson.databind.ObjectMapper

/** Fetches Avro value-schemas from the Schema Registry -- the Scala port of
  * apps/schema_registry.py (PLAN.md 5.3). Java 11's HttpClient and the
  * Jackson that Spark already ships keep this dependency-free.
  */
object SchemaRegistry {

  val Url = "http://schema-registry:8081"

  private val client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build()
  private val mapper = new ObjectMapper()

  /** Fetched once per entity at job startup, not per record: a schema change
    * while running either decodes wrong and raises, or silently omits a new
    * field, until the job restarts (the tradeoff accepted in PLAN.md 3.2).
    */
  def fetchLatestSchema(topic: String): String = {
    val request = HttpRequest
      .newBuilder(URI.create(s"$Url/subjects/$topic-value/versions/latest"))
      .timeout(Duration.ofSeconds(10))
      .GET()
      .build()

    val response = client.send(request, HttpResponse.BodyHandlers.ofString())
    if (response.statusCode() != 200) {
      throw new RuntimeException(
        s"Schema Registry returned ${response.statusCode()} for $topic-value: ${response.body()}")
    }
    mapper.readTree(response.body()).get("schema").asText()
  }
}
